Sources/SwiftDocC/Utility/CollectionChanges.swift

```diff
@@ -0,0 +1,215 @@
+        // Insertions are applied in order. This means that we can start with the previous offset and index.
+        var offset = insertStartOffset
+        for segmentIndex in insertStartIndex ..< segments.count {
+            let segment = segments[segmentIndex]
+            if segment.kind == .remove {
+                continue
+            }
+            
+            if index <= offset + segment.count {
+                return (segment, offset, segmentIndex)
+            }
+            offset += segment.count
+        }
+        return nil
+    }
+    
+    mutating func insert(at insertIndex: Int) {
+        guard let (segment, startOffset, segmentIndex) = findSegment(toInsertAt: insertIndex) else {
+            assert(segments.count == 1 && segments[0].kind == .remove, """
+                The only case when a segment can't be found in the loop is if the only segment is a 'remove' segment.
+                This happens when all the 'common' elements are removed (meaning that the 'from' and 'to' values have nothing in common.
+                """)
+            
+            segments.append(Segment(kind: .insert, count: 1))
+            return
+        }
+        assert(segment.kind != .remove)
+        
+        insertStartOffset = startOffset
+        insertStartIndex  = segmentIndex
+        
+        guard segment.kind != .insert else {
+            segments[segmentIndex].count += 1
+            return
+        }
+        assert(segment.kind == .common)
+        
+        if insertIndex == startOffset {
+            // Insert at start of segment
+            segments.insert(Segment(kind: .insert, count: 1), at: segmentIndex)
+        } else if insertIndex == startOffset + segment.count {
+            // Insert at end of segment
+            let insertSegmentIndex = segmentIndex + 1
+            
+            // If this is the last segment, append a new 'insert' segment
+            guard insertSegmentIndex < segments.count else {
+                segments.append(Segment(kind: .insert, count: 1))
+                return
+            }
+            
+            switch segments[insertSegmentIndex].kind {
+            case .insert:
+                // If the next segment is an 'insert' segment, simply increment it
+                segments[insertSegmentIndex].count += 1
+                
+            case .common:
+                // If the next segment is a 'common' segment, insert a new 'insert' segment before it
+                segments.insert(Segment(kind: .insert, count: 1), at: insertSegmentIndex)
+                
+            case .remove:
+                // If the next segment is a 'remove' segment, skip over it so that insertions are always after removals.
+                segments.insert(Segment(kind: .insert, count: 1), at: insertSegmentIndex + 1)
+                
+                assert(insertSegmentIndex + 2 == segments.count || segments[insertSegmentIndex + 2].kind == .common,
```

Comment:
This essentially just asserts that insertions are ordered from lowest index to highest.

Reply:
Maybe if we want to make that kind of assertion, we could assert that the collection difference traversal ordered is correct. That way if that ever changes we can get a clear signal and fix the code.